### PR TITLE
[JENKINS-63587] Allow disabling Ant default excludes in the zip step

### PR DESCRIPTION
**Pull request by ykhandelwal913**
Bydefault zipcreation excludes the default files like .gitignore and others mentioned  in https://ant.apache.org/manual/dirtasks.html#defaultexcludes.

This PR is to solve https://issues.jenkins.io/browse/JENKINS-63587

Example:
`zip zipFile: 'template.zip', archive: true, defaultExcludes:false, glob: 'template-*/**/*'`
default value of defaultExcludes is true.

### CODE REVIEW

**Review by rsandell — APPROVED** [review]

**Review by rsandell — CHANGES_REQUESTED** [review]
Missing documentation

**Comment by rsandell** [comment]
See https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep and https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/tar/TarStep

Also if this only applies to zip files then the parameter should be added to that step, not the abstract step that also is used by `TarStep`.

**Comment by ykhandelwal913** [comment]
> See https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep and https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/tar/TarStep
> 
> Also if this only applies to zip files then the parameter should be added to that step, not the abstract step that also is used by `TarStep`.

It is only applies to zip files.

**Comment by ykhandelwal913** [comment]
> Missing documentation

Done. Added in code itself. is that fine?

**Comment by rsandell** [comment]
You also need to add a `help-defaultExcludes.html` to the resource directory https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep with user documentation explaining what it is about, not every user can read java code :)

**Comment by ykhandelwal913** [comment]
@rsandell not sure why build is failing. Can you help?

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-defaultExclude.html
Comment (rsandell) [inline src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-defaultExclude.html:26]: This doesn't seem to be correct. It's a checkbox, how is the user supposed to provide an ant style pattern?

**Review by rsandell — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java:124]: Because this feature only applies to Zip files, and the configuration options is only in that class, this code should be moved into the `ZipStep` class.

**Review by rsandell — CHANGES_REQUESTED** [review]

**Comment by basil** [comment]
Looks like the build failure is caused by a Javadoc error:

```
16:49:16  [ERROR] /home/jenkins/agent/workspace/line-utility-steps-plugin_PR-180/src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java:132: error: invalid use of @return
16:49:16  [ERROR]      * @return the defaultExcludes boolean
16:49:16  [ERROR]        ^
```

**Comment by basil** [comment]
PS You can test Javadoc generation locally by running `mvn clean verify -DskipTests -Pjenkins-release`.

**Comment by ykhandelwal913** [comment]
> Looks like the build failure is caused by a Javadoc error:
> 
> ```
> 16:49:16  [ERROR] /home/jenkins/agent/workspace/line-utility-steps-plugin_PR-180/src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java:132: error: invalid use of @return
> 16:49:16  [ERROR]      * @return the defaultExcludes boolean
> 16:49:16  [ERROR]        ^
> ```

Thanks for the pointer. Fixed it.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java
Comment (ykhandelwal913) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java]: Moved it to ZipStep class. Can you review please.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java
Comment (basil) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:78]: Making this field `final` would make it both consistent with the other fields in this class and easier to reason about (mutable state is generally harder to reason about than immutable state).

Comment (ykhandelwal913) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:78]: @basil If i make it final, how can i assign the default value as true as by default we want not to includes patterns which are excludes by ant java pattern.

Comment (ykhandelwal913) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java]: Done. can you review?

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-defaultExclude.html
Comment (basil) [inline src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-defaultExclude.html:28]: This documentation is grammatically confusing (the first sentence probably meant to say "_Whether_ the zip file should […]") and also has incorrect capitalization. Suggest clarifying as follows:

```
<p>By default, the following exclusions are applied:</p>

<pre>**/*~
**/#*#
**/%*%
**/.#*
**/._*
**/.bzr
**/.bzr/**
**/.bzrignore
**/CVS
**/CVS/**
**/.cvsignore
**/.DS_Store
**/.git
**/.git/**
**/.gitattributes
**/.gitignore
**/.gitmodules
**/.hg
**/.hg/**
**/.hgignore
**/.hgsub
**/.hgsubstate
**/.hgtags
**/SCCS
**/SCCS/**
**/.svn
**/.svn/**
**/vssver.scc
</pre>

<p>Set this option to <code>false</code> if you do not want these default exclusions to be applied.</p>
```

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java
Comment (basil) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java:52]: Incorrect: this is a boolean, not a pattern.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java
Comment (basil) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java:62]: Incorrect: this is a boolean, not a pattern.

**Comment by basil** [comment]
> It is only applies to zip files.

From my reading of the code, it appears the tar step uses the same Ant task and therefore the same default exclusions. If true, then it seems inconsistent to add this parameter to the unzip step but not the tar step.

**Comment by ykhandelwal913** [comment]
> From my reading of the code, it appears the tar step uses the same Ant task and therefore the same default exclusions. If true, then it seems inconsistent to add this parameter to the zip step but not the tar step.

As this ticket was for zip step i added code only for that but i agree with your point and to have the same consistency in zip and tar, added the same in tarStep as well.

**Comment by ykhandelwal913** [comment]
@basil can this be reviewed now. This feature is much needed for us to progress in our project. Hence the urgency.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java
Comment (basil) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java:37]: I do not see a reason to add an import that is not used. This seems like leftover code from an earlier revision. Suggest removing the addition of this import.

**Review by basil — COMMENTED** [review]
This is looking good overall, but I think now that the scope of this PR extends to both the zip and the tar step, HTML documentation for the tar step needs to be added to match that of the zip step, along with a corresponding test for the tar step. Otherwise the change is inconsistent, applying some logic to one step but not the other.

**Review by basil — APPROVED** [review]
Looks good to me. Thanks for your contribution and for addressing feedback so quickly! In order to set expectations properly, I am not a maintainer of this plugin (though I am a Jenkins core maintainer), so I will be unable to merge your change. But @rsandell is a maintainer of this plugin, and I believe that the work we have done together has brought this PR closer to a state that can be approved by him (if not to the finish line!). Thanks again!

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java
Comment (ykhandelwal913) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java]: Done.

**Comment by ykhandelwal913** [comment]
> Looks good to me. Thanks for your contribution and for addressing feedback so quickly! In order to set expectations properly, I am not a maintainer of this plugin (though I am a Jenkins core maintainer), so I will be unable to merge your change. But @rsandell is a maintainer of this plugin, and I believe that the work we have done together has brought this PR closer to a state that can be approved by him (if not to the finish line!). Thanks again!

Thanks for the feedback and support provided.

**Comment by rsandell** [comment]
Looks good, from what I can see you only need to fix the javadoc errors in the new code to make the build pass.

**Comment by ykhandelwal913** [comment]
@rsandell can this be merged?

**Comment by ykhandelwal913** [comment]
It didnot create a new release. Do i need to do anything from my side?

**Pull request merged**
